<a href="https://colab.research.google.com/github/wjlee62/Math-381-Wi26/blob/main/Homework_6.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

**Homework 6, due Saturday, February 28**

# Problems 1-5

These problems are to be done by hand.

1. Problem 1, page 927. Define the random variables $X_t$ in the Markov chain, list what the states are, and give the transition probability matrix. In addition, draw a graph represnting the Markov chain.

2. Problem 4, page 927. Same instructions as previous problem. Make sure to answer all parts of the question.

3. Problem 2, page 931.

4. In Problem 1, find the fraction of days it will be sunny.

5. Problem 1, page 957.

# Problem 6

Let $P$ be the transition probability matrix from Problem 3 with $p = 1/2$. Complete the following code snippet to estimate the limit of $P^k$ as $k \to \infty$. It may help to look at coffee.py on the class Github.

In [4]:
import numpy as np
p = 0.5
q = 1-p

P = np.array([[1,0,0,0,0],
 [q, 0, p, 0, 0],
  [0, q, 0, p, 0],
   [0, 0, q, 0, p],
    [0, 0, 0, 0, 1]])

k = 10000

# Q is a matrix. It is an approximation of the limit of P^k as k->infinity.
Q =np.linalg.matrix_power(P, k)

print(Q)

[[1.   0.   0.   0.   0.  ]
 [0.75 0.   0.   0.   0.25]
 [0.5  0.   0.   0.   0.5 ]
 [0.25 0.   0.   0.   0.75]
 [0.   0.   0.   0.   1.  ]]


Interpret the output of your code in terms of the game.

**Interpretation:** In the long run, you cannot have 1, 2, or 3 dollars. You will either eventually lose everything or reach your goal. The game cannot continue forever.  




# Problem 7

Find the stationary distribution of the Markov chain from Problem 2. You should do so by completing the following code. Use the method from coffee.py; **you will not get credit for using another method**.

In [6]:
P = np.array([[0.95,0, 0.05, 0], #SS
              [0.7, 0, 0.3, 0], #CS
              [0, 0.4, 0, 0.6], #SC
              [0, 0.2, 0, 0.8] #CC
              ])

A = np.identity(4) - P.T
A = np.append(A, [[1,1,1,1]], axis = 0)
b = [0,0,0,0,1]
x = np.linalg.lstsq(A,b) [0]

print(x)

[0.73684211 0.05263158 0.05263158 0.15789474]


Using the output of your code, determine the fraction of days it will be sunny.

**Answer and explanation:** According to the output, chances of SS = 0.73684211 and chances of CS = 0.05263158. Hence, the fraction of days it will be sunny = SS + CS = 0.73684211 + 0.05263158 = 0.78947369 (78.9%).